pkg/lightning/mydump/parquet_parser.go

```diff
@@ -272,67 +372,130 @@ func (pp *ParquetParser) ReadRow() error {
 		}
 	default:
 		log.L().Fatal("unknown value", zap.Stringer("kind", v.Kind()),
 			zap.String("type", v.Type().Name()), zap.Reflect("value", v.Interface()))
+		return errors.Errorf("unknown value: %v", v)
+	}
+	return nil
+}
+
+func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
+	if meta.LogicalType != nil && meta.LogicalType.DECIMAL != nil {
+		v = binaryToDecimalStr([]byte(v), int(meta.LogicalType.DECIMAL.Scale))
+	}
+	d.SetString(v, "")
+}
+
+func binaryToDecimalStr(rawBytes []byte, scale int) string {
+	negative := rawBytes[0] > 127
+	if negative {
+		for i := 0; i < len(rawBytes); i++ {
+			rawBytes[i] = ^rawBytes[i]
+		}
+		for i := len(rawBytes) - 1; i >= 0; i-- {
+			rawBytes[i] += 1
+			if rawBytes[i] != 0 {
+				break
+			}
+		}
+	}
+
+	intValue := big.NewInt(0)
+	intValue = intValue.SetBytes(rawBytes)
+	val := fmt.Sprintf("%0*d", scale, intValue)
+	dotIndex := len(val) - scale
+	var res strings.Builder
+	if negative {
+		res.WriteByte('-')
+	}
+	if dotIndex == 0 {
+		res.WriteByte('0')
+	} else {
+		res.Write([]byte(val[:dotIndex]))
 	}
+	if scale > 0 {
+		res.WriteByte('.')
+		res.Write([]byte(val[dotIndex:]))
+	}
+	return res.String()
 }
 
 // when the value type is int32/int64, convert to value to target logical type in tidb
-func setDatumByInt(d *types.Datum, v int64, meta *parquet.SchemaElement) {
-	if meta.ConvertedType == nil {
+func setDatumByInt(d *types.Datum, v int64, meta *parquet.SchemaElement) error {
+	if meta.ConvertedType == nil && meta.LogicalType == nil {
 		d.SetInt64(v)
-		return
+		return nil
 	}
-	switch *meta.ConvertedType {
-	// decimal
-	case parquet.ConvertedType_DECIMAL:
-		minLen := *meta.Scale + 1
+
+	logicalType := meta.LogicalType
+	switch {
+	case logicalType.DECIMAL != nil:
+		if logicalType.DECIMAL.Scale == 0 {
+			d.SetInt64(v)
+			return nil
+		}
+		minLen := logicalType.DECIMAL.Scale + 1
 		if v < 0 {
 			minLen++
 		}
 		val := fmt.Sprintf("%0*d", minLen, v)
 		dotIndex := len(val) - int(*meta.Scale)
 		d.SetString(val[:dotIndex]+"."+val[dotIndex:], "")
-	case parquet.ConvertedType_DATE:
+	case logicalType.DATE != nil:
 		dateStr := time.Unix(v*86400, 0).Format("2006-01-02")
 		d.SetString(dateStr, "")
-	// convert all timestamp types (datetime/timestamp) to string
-	case parquet.ConvertedType_TIMESTAMP_MICROS:
-		dateStr := time.Unix(v/1e6, (v%1e6)*1e3).Format("2006-01-02 15:04:05.999")
-		d.SetString(dateStr, "")
-	case parquet.ConvertedType_TIMESTAMP_MILLIS:
-		dateStr := time.Unix(v/1e3, (v%1e3)*1e6).Format("2006-01-02 15:04:05.999")
+	case logicalType.TIMESTAMP != nil:
+		// convert all timestamp types (datetime/timestamp) to string
+		var sec, nsec int64
+		units := logicalType.TIMESTAMP.Unit
+		if units.MICROS != nil {
+			sec = v / 1e6
+			nsec = (v % 1e6) * 1e3
+		} else if units.MILLIS != nil {
+			sec = v / 1e3
+			nsec = (v % 1e3) * 1e6
+		} else {
+			// nano
+			sec = v / 1e9
+			nsec = v % 1e9
+		}
+		// TODO: how to deal with TimeZone
```

Comment:
Is there more we need to do about time zones?

Reply:
The time/timestamp value read from parquet is either UTC or Local timezone (based on the `IsAdjustedToUTC` setting). Since currently lightning don't support set timezone (always set to target cluster's timezone). Not sure how should we deal with the unknown timezone offset if `IsAdjustedToUTC = false` (at least the default is true). @kennytm  PTAL

Reply:
Make the produced string `2006-01-02 15:04:05.999999Z` when isAdjustedToUTC is true, and `2006-01-02 15:04:05.999999` when it is false.

(why do we only support 3 fractional digits instead of 6?)

Reply:
done